File: src-java/muance/javafx/ListCell.java
package muance.javafx;

import clojure.lang.IFn;

public class ListCell<T> extends javafx.scene.control.ListCell<T> {
  private IFn updateItem;
  private Object state;

  public ListCell(IFn updateItem, Object state) {
    this.updateItem = updateItem;
    this.state = state;
  }

  public ListCell(IFn updateItem) {
    this.updateItem = updateItem;
  }

  @Override
  public void updateItem(T item, boolean empty) {
    super.updateItem(item, empty);
    updateItem.invoke(this, item, empty);
  }

  public Object getState() {
    return state;
  }
}

File: src/muance/list_cell.clj
(ns muance.list-cell
  (:require [muance.arrays :as a]
            [muance.diff :as diff]
            [muance.vtree :as vtree])
  (:import [java.util ArrayList]
           [muance.javafx ListCell]))

(def ^:const index-vtree 0)
(def ^:const index-component 1)

(defn list-cell-updateItem [^ListCell this item empty]
  (if (or (nil? item) empty)
    (.setGraphic this nil)
    (let [vtree-component (.getVtreeComponent this)
          vtree (a/aget vtree-component index-vtree)
          render-queue (vtree/render-queue vtree)
          comp-fn (a/aget vtree-component index-component)
          vnode (vtree/vnode vtree)
          comp (a/aget (vtree/vnode vtree) diff/index-children 0)]
      (diff/patch-impl render-queue vnode comp comp-fn item false)
      (diff/process-post-render-hooks render-queue))))

